src/pcre2_compile.c

```diff
@@ -5178,7 +5187,722 @@ errorcode = ERR79;
+          memctl->free(alt_positions, memctl->memory_data);
+        return;
+      }
+    }
+
+    if (alt_positions != alt_position_buf)
+      memctl->free(alt_positions, memctl->memory_data);
+  }
+
+  /* We didn't rewrite this group
+   * Even so, a rewrite of some subgroups might be needed */
+  DONT_REWRITE: ;
+  uint32_t *item = first;
+  if (buf->start != NULL) {
+    /* It has already been decided that some part of the overall pattern does need to be rewritten */
+    rewrite_buf_copy(buf, start, first - start, memctl);
+    while (item < end) {
+      uint32_t *item_quant = NULL;
+      uint32_t *item_end = scan_item(item, end, &item_quant);
+      uint32_t meta = META_CODE(*item);
+      if (is_group_starter(meta)) {
+        rewrite_alternation(item, item_end, item_quant, buf, pattern, patstring, memctl);
+      } else {
+        rewrite_buf_copy(buf, item, item_end - item, memctl);
+      }
+      item = item_end;
+      REWRITE_UNDER_WAY: ;
+    }
+  } else {
+    /* It hasn't yet been decided that we need to rewrite some part of the overall pattern...
+     * But if we find a subgroup which needs rewrite, then we can still initiate that process */
+    while (item < end) {
+      uint32_t *item_quant = NULL;
+      uint32_t *item_end = scan_item(item, end, &item_quant);
+      uint32_t meta = META_CODE(*item);
+      if (is_group_starter(meta)) {
+        rewrite_alternation(item, item_end, item_quant, buf, pattern, patstring, memctl);
+        if (buf->start != NULL) {
+          /* A rewrite has now started, so go to the other loop */
+          item = item_end;
+          goto REWRITE_UNDER_WAY;
+        }
+      }
+      item = item_end;
+    }
+  }
+}
+
+static void rewrite_regex(compile_block *cb, pcre2_compile_context *ccontext, BOOL heap_parsed_pattern)
+{
+  rewrite_buf buf;
+  rewrite_buf_init(&buf);
+
+  rewrite_alternation(cb->parsed_pattern, cb->parsed_pattern_end-1, NULL, &buf, cb->parsed_pattern, cb->start_pattern, &ccontext->memctl);
+  rewrite_finish(&buf, cb, &ccontext->memctl, heap_parsed_pattern);
+}
+
+/* Strip non-capturing parentheses which may be wrapping the entire regex */
+static void strip_enclosing_nocaptures(compile_block *cb)
+{
+  while (cb->parsed_pattern_end > cb->parsed_pattern && META_CODE(cb->parsed_pattern[0]) == META_NOCAPTURE) {
+    uint32_t *group_end = find_group_end(cb->parsed_pattern, cb->parsed_pattern_end);
+    if (group_end != cb->parsed_pattern_end-1) {
+      break;
+    }
+
+    size_t pattern_length = cb->parsed_pattern_end - cb->parsed_pattern;
+    /* Copy down everything except for unneeded BRA, unneeded KET, and META_END
+     * We will leave some junk in the high end of the parsed_pattern buffer, but
+     * this doesn't matter; `parsed_pattern_end` will point to the end of the used
+     * portion, and we will ensure the used portion ends with META_END */
+    memmove(&cb->parsed_pattern[0], &cb->parsed_pattern[1], (pattern_length - 3) * sizeof(uint32_t));
```

Comment:
wouldn't be easier to override the last record with META_END and move the pointer to the first forward?

Reply:
True, that is very true.

Reply:
Ah, wait a moment, I answered too soon. `cb->parsed_pattern` may, in some cases, be a malloc'd pointer. We have to keep it so that we can `free` it later. So if we want to be able to remove a prefix from the pattern by bumping `cb->parsed_pattern` forward, we would need to add another member to `compile_block` to record the original value.

Do you think that would be worthwhile? Personally, I have my doubts.

Even so, at the very least, I could ensure that `memmove` is only done once even if the pattern is wrapped in multiple layers of redundant parens.

Reply:
> we would need to add another member to `compile_block` to record the original value.

or just keep it in the stack as a local variable, no?; worst case you might need keeping track of the depth and I am sure there is a variable in there that could be reused.

it is not only the performance impact (remember that `compile_block` is in the stack to avoid doing malloc/free and the performance impact that has) of doing a `memmove`, but the fact that it is (most of the time) writing over the stack memory and that is also risky.

Reply:
> or just keep it in the stack as a local variable, no?; worst case you might need keeping track of the depth and I am sure there is a variable in there that could be reused.

That would also work, but we would need to keep that stack variable around all the way to the very end of `pcre2_compile` (currently line 11919-11920). If we do that, I think it's not any different from adding another member to `compile_block`.

Reply:
> I think it's not any different from adding another member to `compile_block`

AFAIK `compile_block` is by design a "context" that gets passed around between possibly recursive calls, and therefore gets updated by them.

Unless I am not understanding your explanation correctly, it would seem that the original pointer to `cb->parsed_pattern` is constant and only would need to be updated in the parent function, no? so wouldn't it make more sense (by design) to keep it as a simple local variable?

Reply:
I'll have to look at the code in detail and see exactly where the original pointer to `cb->parsed_pattern` is needed. Then I can give a better answer.

Reply:
The "compile block", pointed at by the variable cb, contains data that is needed by more than one function in the compiling process, and is more than would be sensible to pass as multiple arguments. In a non-thread-safe world these would probably be static variables, but the PCRE2 functions (with the exception of the POSIX wrapper stuff) are thread-safe and therefore cannot use static variables. If a value is needed only in one function, it can just be on the local stack.

Reply:
@carenas is right 😆 
Updated my PR accordingly. It's better this way.

Reply:
> If a value is needed only in one function, it can just be on the local stack.

That is what I meant by "design", and indeed why I was expecting it to be set as a local variable for pcre2_compile() where its scope is simpler to understand instead of something that _could_ be updated by any function that participates in the process of compiling a pattern.

It seems I wasn't "right" enough though as the implementation went in a different direction.

Reply:
@carenas Yes, I discovered that it wasn't that easy. The problem was that if a rewrite is performed, the buffer which the rewritten pattern goes into *also* needs to be reliably freed.